### PR TITLE
 [ConstraintSystem] Don't allow explicit closure result to be implicitly converted to `Void`

### PR DESCRIPTION
**Pull request by xedin**
It's allowed to convert a single statement closure from `(...) -> T` to `(...) -> Void`
_only_ if there is no explicit `return` in the body.

Resolves: [SR-12277](https://bugs.swift.org/browse/SR-12277)
Resolves: rdar://problem/52204414

### CODE REVIEW

File: include/swift/AST/TypeCheckRequests.h [hunk 1]
```diff
@@ -2141,6 +2142,24 @@ class ScopedImportLookupRequest
   bool isCached() const { return true; }
 };
 
+/// Determine whether closure body has any explicit `return`
+/// statements which could produce a non-void result.
+class ClosureHasExplicitResultRequest
```
Comment (xedin) [inline include/swift/AST/TypeCheckRequests.h]: @hamishknight Did I do this correctly?

**Review by hamishknight — APPROVED** [review]

File: include/swift/AST/TypeCheckRequests.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/TypeCheckRequests.h]: Looks good to me!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by hborla — APPROVED** [review]

**Pull request merged**